### PR TITLE
member: refactor election

### PR DESCRIPTION
**Pull request by JmPotato**
Signed-off-by: JmPotato <ghzpotato@gmail.com>

### What problem does this PR solve?

Refactor election logic.

### What is changed and how it works?

Pull away some common methods of election from `member` and prepare for incoming TSO refactoring next.

### Check List

Tests

- Unit test
- Integration test

Code changes

- Has leader election change

### Release note

Refactor election logic.

### CODE REVIEW

File: server/election/leadership.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+func (ls *Leadership) setLease(lease *Lease) {
+	ls.lease.Store(lease)
+}
+
+// ResetLease sets the lease of leadership to nil.
+func (ls *Leadership) resetLease() {
+	ls.setLease(nil)
+}
+
+// NewLeadership ...
```
Comment (disksing) [inline server/election/leadership.go:73]: ？

File: server/election/leadership.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+func (ls *Leadership) setLease(lease *Lease) {
+	ls.lease.Store(lease)
+}
+
+// ResetLease sets the lease of leadership to nil.
+func (ls *Leadership) resetLease() {
+	ls.setLease(nil)
+}
+
+// NewLeadership creats a new Leadership.
```
Comment (disksing) [inline server/election/leadership.go:73]: creates

File: server/server.go [hunk 3]
```diff
@@ -469,9 +473,10 @@ func (s *Server) LoopContext() context.Context {
 
 func (s *Server) startServerLoop(ctx context.Context) {
 	s.serverLoopCtx, s.serverLoopCancel = context.WithCancel(ctx)
-	s.serverLoopWg.Add(3)
+	s.serverLoopWg.Add(4)
 	go s.leaderLoop()
 	go s.etcdLeaderLoop()
+	go s.tsoAllocatorLoop()
```
Comment (disksing) [inline server/server.go:479]: I think better not to incldue the `tsoAllocator` in this PR.

Comment (JmPotato) [reply server/server.go:479]: The reason I include this change is try to show you some ideas about what the later work will be like in order to help review. I'll take away these changes though for this pr.

File: server/election/lease.go [hunk 4]
```diff
@@ -11,39 +11,50 @@
...
-	ID           clientv3.LeaseID
-	leaseTimeout time.Duration
+const (
+	revokeLeaseTimeout = time.Second
+	requestTimeout     = etcdutil.DefaultRequestTimeout
+	slowRequestTime    = etcdutil.DefaultSlowRequestTime
+)
 
-	expireTime atomic.Value
+// Lease is used for campaigning and renewing elected leadership.
```
Comment (Yisaer) [inline server/election/lease.go:34]: I didn't get how Lease is used for `campaigning` here as it still have the same method comparing to the previous `LeaderLease`. Can you explain why adding `campaigning` here?

File: server/election/lease.go [hunk 5]
```diff
@@ -11,39 +11,50 @@
...
+type Lease struct {
+	// purpose is used to show what this election for
+	Purpose string
+	// etcd client and lease
+	client *clientv3.Client
+	lease  clientv3.Lease
+	ID     clientv3.LeaseID
+	// lifetime related fields
```
Comment (Yisaer) [inline server/election/lease.go:42]: I think the comment here could be more detailed as I didn't get the point for this comment.

File: server/election/leadership.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
+// Copyright 2019 PingCAP, Inc.
```
Comment (Yisaer) [inline server/election/leadership.go:1]: should be 2020.

File: server/election/leadership.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+// NewLeadership creates a new Leadership.
+func NewLeadership(client *clientv3.Client, purpose string) *Leadership {
+	leadership := &Leadership{
+		Purpose: purpose,
+		client:  client,
+	}
+	return leadership
+}
```
Comment (Yisaer) [inline server/election/leadership.go:80]: I suggest place `NewLeadership` function under `Leadership` definition.

File: server/election/lease.go (hunk 4 above)
Comment (JmPotato) [reply server/election/lease.go]: `Lease` is used as the low-level mechanism for campaigning and renewing elected leadership. The way to gain and maintain leadership is to update and keep the lease alive continuously. So this package wraps some `Lease` related methods that `Leadership` will call during the campaigning process.

**Comment by ZenoTan** [comment]
LGTM

**Pull request merged**